triton-cli/src/main.rs

```diff
@@ -0,0 +1,214 @@
+        CliArg::Verify { proof_path } => verify(&proof_path)?,
+    }
+    Ok(())
+}
+
+fn digest_to_str(d: Digest) -> String {
+    let mut hex = String::new();
+    d.0.iter()
+        .for_each(|v| write!(&mut hex, "{:16x}", u64::from(v)).unwrap());
+    format!("0x{hex}")
+}
+
+fn verify(proof_path: &str) -> Result<()> {
+    let (stark, claim, proof) = read_proof(proof_path).expect("Failed to load proof");
+
+    stark
+        .verify(&claim, &proof, &mut None)
+        .with_context(|| "Stark proof verification failed")?;
+
+    println!("proof is valid!");
+    println!("program digest: {}", digest_to_str(claim.program_digest));
+    println!("=================");
+    println!("security level: {} bits", stark.security_level);
+    println!("FRI expansion factor: {}", stark.fri_expansion_factor);
+    println!("trace randomizers: {}", stark.num_trace_randomizers);
+    println!("colinearity checks: {}", stark.num_collinearity_checks);
+    println!("codeword checks: {}", stark.num_combination_codeword_checks);
+    println!("=================");
+    println!("public inputs:");
+    if claim.input.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.input {
+        println!("{v}");
+    }
+    println!("public outputs:");
+    if claim.output.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.output {
+        println!("{v}");
+    }
+    Ok(())
+}
+
+fn parse_inputs(inputs: Option<String>) -> Vec<BFieldElement> {
+    inputs
+        .unwrap_or_default()
+        .split(',')
+        .filter(|v| !v.is_empty())
+        .map(|v| v.parse().unwrap())
+        .collect()
+}
+
+fn prove(
+    asm_filepath: &str,
+    out: &str,
+    public_inputs: Option<String>,
+    private_inputs: Option<String>,
+) -> Result<()> {
+    if std::path::Path::new(out).exists() {
+        anyhow::bail!("output file already exists: {out}");
+    }
+    let asm = fs::read_to_string(asm_filepath)
+        .with_context(|| "Failed to read Triton assembly from file")?;
+    let program = triton_program!({ asm });
+
+    let public_input = PublicInput::from(parse_inputs(public_inputs));
+    let non_determinism = NonDeterminism::from(parse_inputs(private_inputs));
+    println!("proving...");
+    let data = triton_vm::prove_program(&program, public_input, non_determinism)
+        .with_context(|| "Triton VM errored during program execution")?;
+    println!("success!");
+
+    write_proof(data, out).with_context(|| "Failed to write proof to file")?;
+    println!("proof written to: {out}");
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct SerializedProof {
+    version: u8,
+    security_level: u64,
+    fri_expansion_factor: u64,
+    num_trace_randomizers: u64,
+    num_colinearity_checks: u64,
+    num_combination_codeword_checks: u64,
+    program_digest: Vec<u64>,
+    public_inputs: Vec<u64>,
+    public_outputs: Vec<u64>,
+    proof: Vec<u64>,
+}
```

Comment:
Getting this right seems to be more difficult (for me) than I had anticipated. I see more or less three main ways we can go about it:

1. Declare proof versioning out of scope for Triton CLI.
2. Detect out-of-sync versions and report them as an error.
3. Support as many proof versions as possible.

## 1 – Declare Versioning to be Out of Scope

The simplest “solution” from an implementation standpoint. If deserialization fails, the cause might be an out-of-date proof, hard disk failure, aliens, …. The user can check for an update of Triton CLI and try their luck again.

## 2 – Error on Out-of-Sync Proof

Requires
- storing the proof's version alongside the proof, and
- knowing the version of Triton VM.

## 3 - Support Different Proof Versions

Requires everything from (2) plus old versions of Triton VM as additional dependencies. Then, proof generation / verification can be delegated to the chosen version. This is certainly possible by [renaming dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml). The biggest downside I see are compilation times and binary size – Triton VM is not a particularly leightweight library.

# Detecting Versions

## Of the Proof

For strategies 2 and 3, the proof's version must be accessible whether deserialization works or not. Here are suggestions that might work:
1. Use the file's first few bytes for the proof's version, store the serialized proof in `bytes[some_index..]`.
2. Declare a big `enum` with one struct variant per proof version. For example:
```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
enum VersionedSerializableProof {
	V0_13 { stark: triton_vm_0_13::Stark, claim:  triton_vm_0_13::Claim, .. },
	V0_22 { stark: triton_vm_0_22::Stark, claim:  triton_vm_0_22::Claim, .. },
	..
}
```
I'm not sure this works, though: I don't think `serde` is able to distinguish between two different versions of same type.

## Of Triton CLI / VM

With strategy (2) “Error on out-of-sync proof”, the version of Triton VM in use must be known. The ways I see to achieve this, are:
1. Use compile-time environment variable `CARGO_PKG_VERSION` via the `env!` macro to get a `&'static str` to the version specifier in the `Cargo.toml` of Triton CLI. Only works under the assumption that Triton CLI has the same version as Triton VM. Notably, this is not currently the case.
2. Use a `build.rs` build script and some parsing of the `Cargo.toml` to populate a compile-time environment variable with the version of Triton VM in use, which can then be read via the `env!` macro.
3. Manually update some hardcoded `const` whenever the version of Triton VM is bumped.

I must admit, none of these approaches seem particularly resilient.


What do you think?  Do these thoughts and conclusions make sense? Am I overlooking something blatantly obvious?

Reply:
In versioning we have 2 things that can change. The first is the structure of the binary file itself. This affects how we implement encoding/decoding, but does not necessarily affect how the proof is verified. The second is the version of triton-vm that generated the proof. This is much more difficult to handle because a proof might be (de)serialized just fine and then fail verification because e.g. the merkle structure was off by 1 or the hash function changed etc.

The first case is firmly in scope for this package. Anytime we change the `SerializedProof` structure we should keep the old structure/implementation somewhere and `match` into the old structure based on the version indicator until we explicitly drop support for a version. The second case seems mostly out of scope for this package. By that i mean, we should not try and support anything but the current version. The triton-cli version number should follow triton-vm. e.g. anytime a breaking version of triton-vm is released, a breaking triton-cli version should be released as well. I would lean into the versioning system here and use the `CARGO_PKG_VERSION` to throw an error if a proof is verified from the wrong major version.

The tricky part here is _detecting_ when a breaking change is made to the STARK prover/verifier logic and strictly following versioning rules. I would commit a proof to this repo from the current version of triton-vm and have the CI attempt to verify this proof on every push. This will fail anytime a breaking change is made which is an indicator to push a breaking version release (or try to undo the breaking change). Sort of a makeshift e2e test.

I think to mitigate some of the risk here we can add more variables into the serialized proof. Namely:
- name/identifier of hash function used
- field/field extension used

This would add slightly more flexibility on detecting implementation details.

Reply:
I would handle this versioning stuff lazily though. Let's call these things the "proof file version" and the "prover version". 

We start with proof file version 1, there are no other versions to support right now. We start with prover version `0.x.x`. We specify that if the prover version is not defined in the proof file it is assumed to be `0.x.x`.

To _pedantically_ support this I think we need to add the `CARGO_PKG_VERSION` check logic to the prover/verification and throw an error on prover version `>0.x.x`. It would be good to add a comment near this logic documenting this conversation/plan.

Reply:
I applied this in [553a028](https://github.com/TritonVM/triton-vm/pull/305/commits/553a028b40c14e1a4c6aa7036adf9b7b8aa4c3d0).

I ended up putting the version in the proof file as it seemed more self-documenting.

Reply:
BTW i don't know how rust versioning works. I'm using the [semver](https://semver.org/) mental model.